upstream/upkubeless/provision.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package upkubeless
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+
+	"github.com/hashicorp/go-getter"
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-plugin-apps/apps"
+	"github.com/mattermost/mattermost-plugin-apps/utils"
+)
+
+// ProvisionApp creates all necessary functions in Kubeless, and outputs
+// the manifest to use.
+//
+// Its input is a zip file containing:
+//   |-- manifest.json
+//   |-- function files referenced in manifest.json...
+func ProvisionApp(bundlePath string, log utils.Logger, shouldUpdate bool) (*apps.Manifest, error) {
+	dir, err := os.MkdirTemp("", "")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create temp directory to unpack the bundle")
+	}
+	defer os.RemoveAll(dir)
+
+	pwd, err := os.Getwd()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to obtain current working directory")
+	}
+
+	getBundle := getter.Client{
+		Mode: getter.ClientModeDir,
+		Src:  bundlePath,
+		Dst:  dir,
+		Pwd:  pwd,
+		Ctx:  context.Background(),
+	}
+	err = getBundle.Get()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get bundle "+bundlePath)
+	}
+
+	data, err := os.ReadFile(filepath.Join(dir, "manifest.json"))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to load manifest.json")
+	}
+	m, err := apps.ManifestFromJSON(data)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid manifest.json")
+	}
+	log.Debugw("Loaded App bundle",
+		"bundle", bundlePath,
+		"app_id", m.AppID)
+
+	kubelessPath, err := exec.LookPath("kubeless")
```

Comment:
Non blocking for this PR: Given that `kubeless` is written in Go I'm wondering if we could import the code directly instead of requiring the cli command.

Reply:
I tried, I was not very successful.

Reply:
Okay. Is it worth creating an investigation ticket?

Reply:
Discussed offline, likely not worth it.